### PR TITLE
SANS - Add Process All Button

### LINKED ISSUE #23795: Add process selected to SANS GUI
Currently process will process only selected rows or all rows if no rows are selected. After discussion with the scientists it was decided to add two buttons process all and process selected to represent this functionality in a less confusing manner.

### PR DESCRIPTION
**Pull request by TWJubb**
### Summary

This PR adds a process all button to the SANS interface; causing the processing of all non-empty rows in the table. The "Process" button has been changed to "Process Selected" and only processes the rows from the table which are currently selected.

The process all button should appear only when there are non-empty rows in the table, otherwise it should be disabled.

I have also done some minor refactoring in the model view and presenter code of the front tab of the SANS interface.

Additionally cleans up the instrument parser code as implemented in PR #24098 

### Testing

To test, load a masking file and a batch file into the GUI (Interfaces->SANS 2); process all should be available once loaded. Experiment with this and check that the buttons process the rows correctly.

Fixes #23795 

--------------------

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by Matthew-Andrew — CHANGES_REQUESTED** [review]
Code changes are good and in many cases improve things. The only issue I found was that the Process All button is not currently disabled during a process All reduction. This is presumably because of the additional logic to disable it if there are no valid rows which I'm not sure is necessary. Also are there any unit tests which it would be relevant to add here?

**Comment by TTitcombe** [comment]
@Mantid-Matthew this is ready for review now

**Review by Matthew-Andrew — DISMISSED** [review]
The process all button now remains greyed out until a reduction is finished. Good Job

**Comment by Mantid-Matthew** [comment]
@TTitcombe There are now some merge conflicts to resolve as predicted however.

**Review by Matthew-Andrew — DISMISSED** [review]
The process all button now remains greyed out until a reduction is finished. Re approving as review dismissed due to rebase.

**Review by Matthew-Andrew — DISMISSED** [review]
Assuming tests pass this now looks good.

**Review by Matthew-Andrew — DISMISSED** [review]

File: scripts/SANS/sans/gui_logic/models/table_model.py [hunk 1]
```diff
@@ -309,6 +309,10 @@ def to_list(self):
                 self.options_column_model.get_options_string()]
 
     def _convert_sample_shape_to_string(self, shape):
+        if isinstance(shape, str):
+            # TODO temporary fix to shape already being a str
+            # find out where this is being converted as we are trying to convert twice
+            return shape
```
Comment (soininen) [inline scripts/SANS/sans/gui_logic/models/table_model.py]: Is there an issue for this? Would it be worth opening one?

File: scripts/SANS/sans/gui_logic/presenter/run_tab_presenter.py [hunk 2]
```diff
@@ -477,35 +553,49 @@ def notify_progress(self, row, out_shift_factors, out_scale_factors):
...
+    # ----------------------------------------------------------------------------------------------
+    # Row manipulation
+    # ----------------------------------------------------------------------------------------------
+
+    def num_rows(self):
+        return self._table_model.get_number_of_rows()
+
     def on_row_inserted(self, index, row):
+        """
+
```
Comment (soininen) [inline scripts/SANS/sans/gui_logic/presenter/run_tab_presenter.py]: Why no docstring?

**Review by soininen — CHANGES_REQUESTED** [review]
A few questions about commented-out code and other small stuff.

File: scripts/SANS/sans/gui_logic/models/table_model.py (hunk 1 above)
Comment (TTitcombe) [reply scripts/SANS/sans/gui_logic/models/table_model.py]: #24261 now open

**Review by soininen — APPROVED** [review]

**Pull request merged**